### PR TITLE
[vecmem] new package for optional dependency of acts

### PR DESCRIPTION
**Pull request by wdconinc**
This PR cherry-picks some parts of #26790 to make that PR more manageable.

The vecmem package is an optional dependency of acts. This turns it into an independent package in preparation for including it as a dependency of acts.

Maintainer tag: @HadrienG2

### CODE REVIEW

**Comment by wdconinc** [comment]
This probably should inherit from mix-in CudaPackage.

Currently works for `+cuda` and `+hip` but not yet for `+sycl`.

File: var/spack/repos/builtin/packages/vecmem/package.py
Comment (krasznaa) [inline var/spack/repos/builtin/packages/vecmem/package.py]: I'm a bit surprised about the low version number required from CMake. 3.8 would not be able to build even just the core library of the package.

https://github.com/acts-project/vecmem/blob/main/CMakeLists.txt#L8

But with CUDA enabled, the CMake requirement increases even further. (https://github.com/acts-project/vecmem/blob/main/cuda/CMakeLists.txt#L8)

I'm not convinced that spack should code that much information about the project's CMake dependence into its build recipe. But the minimum requirement should really be set to a very minimum of 3.10. Or if you just want to make sure that every part of vecmem can be built, to 3.17.

Comment (HadrienG2) [reply var/spack/repos/builtin/packages/vecmem/package.py]: > I'm not convinced that spack should code that much information about the project's CMake dependence into its build recipe. But the minimum requirement should really be set to a very minimum of 3.10. Or if you just want to make sure that every part of vecmem can be built, to 3.17.

Spack needs to, because it is a full-stack package manager that will install (almost) every dependency including cmake for you. Solving the dependency graph requires knowing which version of each dependency is suitable for installation.

Comment (krasznaa) [reply var/spack/repos/builtin/packages/vecmem/package.py]: I just meant that spack should not need to know that:
  - the absolute minimum is 3.10;
  - if you turn on testing, the minimum becomes 3.11;
  - if you turn on CUDA, the minimum becomes 3.17.

I'm on board with it having one minimum. Which I guess should be 3.17, to be on the safe side...

Comment (HadrienG2) [reply var/spack/repos/builtin/packages/vecmem/package.py]: Ah ok, I thought you meant spack should not know about cmake version requirements. Then we agree. I would also just specify `cmake@3.17:` because it's usually enough and people can still add the full story later on if they ever need a package that only builds with old cmake (which has not happened yet thankfully).

File: var/spack/repos/builtin/packages/vecmem/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/vecmem/package.py]:
```suggestion
    depends_on('cmake@3.17:', type='build')
```

File: var/spack/repos/builtin/packages/vecmem/package.py
Comment (wdconinc) [reply var/spack/repos/builtin/packages/vecmem/package.py]: Implemented.

**Comment by wdconinc** [comment]
With these last changes, this now builds as `vecmem +cuda cuda_arch=50 +hip +sycl ^hipsycl@develop`, when supporting `hipsycl@develop` and adding dependent build environment variables (not included in this branch):
```diff
diff --git a/var/spack/repos/builtin/packages/hipsycl/package.py b/var/spack/repos/builtin/packages/hipsycl/package.py
index d6b069becf..cb146a2a08 100644
--- a/var/spack/repos/builtin/packages/hipsycl/package.py
+++ b/var/spack/repos/builtin/packages/hipsycl/package.py
@@ -23,6 +23,7 @@ class Hipsycl(CMakePackage):
 
     provides("sycl")
 
+    version("develop", branch="develop", submodules=True)
     version("stable", branch="stable", submodules=True)
     version(
         "0.9.1",
@@ -169,3 +170,7 @@ def filter_config_file(self):
         # Replace the installed config file
         with open(config_file_path, "w") as f:
             json.dump(config, f, indent=2)
+
+    def setup_dependent_build_environment(self, env, dependent_spec):
+        env.set('SYCLCXX', self.prefix.bin.syclcc)
+        env.set('SYCLFLAGS', '-I' + self.prefix.include)
```

**Review by HadrienG2 — APPROVED** [review]
Congrats! I'd just wait for feedback from a spack dev on what they consider the best way to propagate SYCL compiler information to dependent builds before merging this (the SYCLCXX mechanism is specific to the vecmem cmake infrastructure, there is unfortunately no standardized support for SYCL builds yet in upstream CMake), but on my side this looks good to go.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
